onnx/defs/shape_inference.h

```diff
@@ -74,11 +90,37 @@ struct InferenceContext {
   virtual const SparseTensorProto* getInputSparseData(size_t index) const = 0;
 };
 
+// We use data propagation to perform partial evaluation of the model, to compute statically
+// known information about tensor values. It is intended to improve the precision of shape
+// inference. We reuse TensorShapeProto to represent the statically known values. One
+// limitation of this is that TensorShapeProto can represent only integer values.
+// As an example, data-propagation is intended to handle code-fragments like below:
+//   shape = Shape(X)
+//   batchsize = Slice(shape, [0], [1])
+//   newshape = Concat (batchsize, [1024, 1024])
+//   Z = Reshape(Y, newshape)
+// If the shape of X is statically known, then data-propagation should be able to determine
+// the value of newshape, as well as the shape of Z.
+struct DataPropagationContext {
+  virtual const AttributeProto* getAttribute(const std::string& name) const = 0;
+  virtual size_t getNumInputs() const = 0;
+  virtual const TypeProto* getInputType(size_t index) const = 0;
+  virtual size_t getNumOutputs() const = 0;
+  virtual TypeProto* getOutputType(size_t index) = 0;
+  virtual ~DataPropagationContext() {}
+  virtual const TensorShapeProto* getInputData(size_t index) = 0;
```

Comment:
Why is the ```const``` missing in ```getOutputType``` and ```getInputData``` ?

Reply:
Current getInputData will add initializer into generatedShapeData_.
getOutputType returns the reference of &allOutputTypes_.

So both of them will potentially modify its member.

Reply:
Can ```getOutputType``` return a ```const TypeProto*```? If so, it can be marked ```const```, right?

Reply:
Yes for `DataPropagationContext`'s `getOutputType`, it is doable. I have updated it as a const function. Thanks. 

Just FYI: `getOutputType` from `InferenceContext` is not a const function either and there are quite many places directly using getOutputType and then modify it (like Loop op or shape_inference.h)...